core/src/server/mod.rs

```diff
@@ -0,0 +1,324 @@
+use crate::repo;
+use axum::{
+    body::StreamBody,
+    http::{header, HeaderMap, StatusCode},
+    response::IntoResponse,
+    routing::get,
+    Extension, Json, Router,
+};
+use flate2::write::GzEncoder;
+use flate2::Compression;
+use serde::{Deserialize, Serialize};
+use simple_logger;
+use std::fs::File;
+use std::sync::{
+    mpsc::{self, Receiver, Sender},
+    Arc, Mutex,
+};
+use tar::Builder;
+use tokio_util::io::ReaderStream;
+
+struct Server {
+    pub port: String,
+    pub tx: Mutex<Sender<Arc<Task>>>,
+    pub rx: Mutex<Receiver<Arc<Task>>>,
+}
+
+type ScanResult = (String, String, repo::Repo);
+
+struct Task {
+    pub task: fn(conf: Arc<crate::config::Config>, repo: Arc<String>) -> Result<ScanResult, String>,
+    pub repo: Arc<String>,
+    pub conf: Arc<crate::config::Config>,
+    //TODO: create type for the Sender and Receiver,
+    //it is quite complicated to read at the moment
+    pub tx: Mutex<Sender<Result<ScanResult, String>>>,
+    pub rx: Mutex<Receiver<Result<ScanResult, String>>>,
+}
+
+impl Task {
+    pub fn new(
+        task: fn(conf: Arc<crate::config::Config>, repo: Arc<String>) -> Result<ScanResult, String>,
+        repo: Arc<String>,
+        conf: Arc<crate::config::Config>,
+    ) -> Self {
+        let (tx, rx) = mpsc::channel();
+        Self {
+            task,
+            repo,
+            conf,
+            tx: Mutex::new(tx),
+            rx: Mutex::new(rx),
+        }
+    }
+}
+
+pub fn run(port: String) {
+    //Initialise the verbose logger
+    //TODO: it should be a little less verbose
+    simple_logger::init().unwrap();
+
+    let (tx, rx) = mpsc::channel();
+    let server: Arc<Server> = Arc::new(Server {
+        port,
+        tx: Mutex::new(tx),
+        rx: Mutex::new(rx),
+    });
+
+    //Start the task scheduler.
+    //
+    //This is a temporary solution, it only allow us to spin up
+    //long running scan operations concurrently without
+    //prioritizing short running operations. Meaning requests
+    //on small repositories will end up sharing the same thread pool
+    //as requests made on big repositories as `kubernetes/kubernetes` and `torvalds/linux`.
+    //A single long running request can slow down all other requests.
+    //
+    //We need to find a way to prioritize short running requests, maybe by running
+    //the long running tasks in a separate thread pool or in separate processes.
+    //
+    //The current solution is good enough for now, but it is not ideal.
+    let s = server.clone();
+    rayon::spawn(move || {
+        while let Ok(task) = s.rx.lock().unwrap().recv() {
+            rayon::spawn(move || {
+                let t = task.clone();
+                let ret = t.task.to_owned()(t.conf.to_owned(), t.repo.to_owned());
+                task.tx.lock().unwrap().send(ret).unwrap();
+            })
+        }
+    });
```

Comment:
This long comment mainly point out the fact that, the way we are handling long running tasks is not ideal, and should be improved later. But it at least give us a solid but slow "non blocking" way of handling multiple requests.
When a user request data from a big repository things get a little saltier.